src/context/QuizContextProvider.jsx

```diff
@@ -0,0 +1,167 @@
+import { useEffect, useState } from 'react';
+import QuizContext from './QuizContext';
+
+const initialState = {
+  quizSettings: {
+    numberOfQuestions: 5,
+    category: 'any',
+    difficulty: 'any',
+  },
+  currentQuestion: 0,
+  questions: [],
+  userAnswers: [],
+  score: 0,
+  cumulativeScore: 0,
+  timer: 10,
+  quizStatus: 'notStarted',
+  categories: [],
+};
+
+export const QuizContextProvider = ({ children }) => {
+  const [quizState, setQuizState] = useState(initialState);
+
+  useEffect(() => {
+    // Fetch categories on mount
+    fetch('https://opentdb.com/api_category.php')
+      .then((response) => response.json())
+      .then((data) => {
+        setQuizState((previousState) => ({
+          ...previousState,
+          categories: data.trivia_categories,
+        }));
+      })
+      .catch((error) => console.error('Error fetching categories:', error));
+  }, []);
+
+  const setQuizSettings = (settings) => {
+    return new Promise((resolve) => {
+      setQuizState((previousState) => {
+        const newState = {
+          ...previousState,
+          quizSettings: { ...previousState.quizSettings, ...settings },
+        };
+        resolve(newState);
+        return newState;
+      });
+    });
+  };
+
+  const handleAnswerQuestion = (answer) => {
+    const isCorrect =
+        answer === quizState.questions[quizState.currentQuestion].correct_answer;
+    setQuizState((previousState) => ({
+      ...previousState,
+      userAnswers: [...previousState.userAnswers, answer],
+      score: isCorrect ? previousState.score + 1 : previousState.score,
+      cumulativeScore: isCorrect ? previousState.cumulativeScore + 1 : previousState.cumulativeScore,
+      currentQuestion: previousState.currentQuestion + 1,
+      timer: 10,
+    }));
+  };
+
+  useEffect(() => {
+    let timer;
+    if (quizState.quizStatus === 'inProgress' && quizState.timer > 0) {
+      timer = setInterval(() => {
+        setQuizState((previousState) => ({
+          ...previousState,
+          timer: previousState.timer - 1,
+        }));
+      }, 1000);
+    } else if (quizState.timer === 0 && quizState.quizStatus === 'inProgress') {
+      handleAnswerQuestion(null);
+      if (quizState.currentQuestion === quizState.questions.length) {
+        handleCompleteQuiz();
+      }
+    }
+    return () => clearInterval(timer);
+  }, [
+    quizState.timer,
+    quizState.quizStatus,
+    quizState.currentQuestion,
+    quizState.questions.length,
+  ]);
```

Comment:
Create a separate custom hook for that and use it in this provider.. Avoid using `let`.